Refactor WithAttackAnimation to support multiple sprite bodies

reaperrr opened the pull request:
Same approach as with `WithMoveAnimation`.

Additionally, reduced string.NullOrEmpty look-ups and made sure the aim/reload sequences don't override a playing attack anim.

atlimit8 commented on OpenRA.Mods.Common/Traits/Render/WithAttackAnimation.cs:
```diff
@@ -88,19 +103,29 @@ void ITick.Tick(Actor self)
 
-			var prefix = (armament.IsReloading && !string.IsNullOrEmpty(info.ReloadPrefix)) ? info.ReloadPrefix : "";
+			var enabledBodies = wsbs.Where(Exts.IsTraitEnabled);
+			if (!enabledBodies.Any())
```
This early abort probably does not even abort early since the `foreach` loop would find nothing and end without doing anything anyways.

reaperrr commented:
Updated.

atlimit8 commented on OpenRA.Mods.Common/Traits/Render/WithAttackAnimation.cs:
```diff
@@ -42,23 +45,35 @@ public class WithAttackAnimation : ITick, INotifyAttack
+				attackAnimPlaying = true;
+				foreach (var wsb in enabledBodies)
+					wsb.PlayCustomAnimation(self, info.AttackSequence,
+						() => { wsb.CancelCustomAnimation(self); attackAnimPlaying = false; });
```
This could cause a problem if the attack animations are different lengths. Perhaps `bool attackAnimPlaying` should be replace with `int attackAnimsPlaying` which would be set to `enabledBodies.Count()` and decremented with `() => { wsb.CancelCustomAnimation(self); attackAnimsPlaying--; }`

pchote replied:
The behaviour here is subtle: right now if any animation is completed then it will restart *all* the animations, which IMO is more consistent than requiring *all* the animations to complete to restart *any*.

Option #3 is to maintain a `HashSet<WithSpriteBody>` or `bool[]` so we know exactly which animations to restart.  I'm not sure whether this is better ingame behaviour than @reaperrr's current implementation, however.

pchote replied:
Do we have a valid usecase for having multiple sprite bodies?  If not, then I think the current code hits the right compromise between simplicity and breaking the least-badly if someone messes up their SpriteBody conditions.

reaperrr replied:
Multiple sprite bodies enabled at the same time might happen in 3rd-party mods, but I think it's extremely unlikely they'd all have attack sequences (and even less likely with varying sequence length).

GraionDilach replied:
I wouldn't assume 3rd party mods would go with multiple enabled-at-the-same-time spritebodies either, they would prolly idleoverlay such cases instead. (Multiple-but-only-one-enabled-at-a-time needs to be valid though for transformer SHP units.)

reaperrr replied:
In that case, I propose the following:
I could change `BodyNames[]` to `BodyName`, looking for `FirstOrDefault` WSB where name matches and trait is enabled.
Additionally, I would then make `WithAttackAnimation` itself conditional, so if someone really wants a 2nd attack anim for a 2nd body active at the same time for some reason, they have the option to use a different name for the 2nd body + 2nd WAA assigned to it.

atlimit8 replied:
That would be the most flexible.

pchote commented on OpenRA.Mods.Common/Traits/Render/WithAttackAnimation.cs:
```diff
@@ -42,23 +45,35 @@ public class WithAttackAnimation : ITick, INotifyAttack
 		void PlayAttackAnimation(Actor self)
 		{
 			if (!string.IsNullOrEmpty(info.AttackSequence))
-				wsb.PlayCustomAnimation(self, info.AttackSequence, () => wsb.CancelCustomAnimation(self));
```
Style nit: it would be more consistent with the check below and our wider code style to invert this if and return early.

reaperrr replied:
Wouldn't a `return` here also make Tick return early even if there's actually an aim- or reload sequence, or would that `return` actually be limited to this method and not bleed through?

Returning if there's no matching body wouldn't be a problem either way, because in that case the aim/reload wouldn't play, either.

That's the reasoning why I did it this way, in any case.

pchote reviewed (commented):
Looks good otherwise, and works as advertised. I'll hold my :+1: until we come to a decision about how to handle multiple bodies.

reaperrr committed:
Refactor WithAttackAnimation

- made trait compatible with actors that have more than one sprite body or enable/disable sprite bodies via conditions
- added check for running attack anim and prevent aim/reload sequences from overriding it
- added caching of whether trait has either aim or reload sequence, to avoid some string.IsNullOrEmpty look-ups every tick

reaperrr committed:
Make WithAttackAnimation conditional

reaperrr commented:
Updated. Travis doesn't seem to trigger on force-pushes anymore, but it passed `make all` and `make check` locally.

pchote commented:
This seems like a more reasonable approach, but IMO we should go further: require all `WithAttackAnimation` traits (and then as a second step all `With*Animation` to define the `WithSpriteBody` that they act on, and then throw a `YamlException` from `RulesetLoaded` if there isn't exactly one matching trait.

reaperrr committed:
Throw yaml exception if WithAttackAnimation has no assigned sprite body

...or too many assigned bodies.

Also, further simplify WithAttackAnimation code.

reaperrr commented:
> require all `WithAttackAnimation` traits (and then as a second step all `With*Animation`) to define the `WithSpriteBody` that they act on, and then throw a `YamlException` from `RulesetLoaded` if there isn't exactly one matching trait.

Updated.

atlimit8 commented on OpenRA.Mods.Common/Traits/Render/WithAttackAnimation.cs:
```diff
@@ -34,68 +34,90 @@ public class WithAttackAnimationInfo : ITraitInfo, Requires<WithSpriteBodyInfo>,
+			if (IsTraitDisabled || noAimOrReloadAnim || attackAnimPlaying)
+				return;
+
+			if (wsb.IsTraitDisabled)
```
I'm on the fence about whether this should be included above. `IsTraitDisabled || noAimOrReloadAnim || attackAnimPlaying || wsb.IsTraitDisabled`

reaperrr replied:
Well, it would reduce the number of lines a bit.

atlimit8 commented on OpenRA.Mods.Common/Traits/Render/WithAttackAnimation.cs:
```diff
@@ -34,68 +34,90 @@ public class WithAttackAnimationInfo : ITraitInfo, Requires<WithSpriteBodyInfo>,
+
+		public override void RulesetLoaded(Ruleset rules, ActorInfo ai)
+		{
+			var matches = ai.TraitInfos<WithSpriteBodyInfo>().Where(w => BodyName == w.Name);
```
`.Count(w => BodyName == w.Name)` can be used instead of `.Where(w => BodyName == w.Name)`, removing an unecessary intermediate enumerable.

reaperrr commented:
Updated.

rob-v commented on OpenRA.Mods.Common/Traits/Render/WithAttackAnimation.cs:
```diff
@@ -34,68 +34,87 @@ public class WithAttackAnimationInfo : ITraitInfo, Requires<WithSpriteBodyInfo>,
-				.Single(a => a.Info.Name == info.Armament);
-			wsb = init.Self.Trait<WithSpriteBody>();
+				.Single(a => a.Info.Name == Info.Armament);
+			wsb = init.Self.TraitsImplementing<WithSpriteBody>().First(w => Info.BodyName == w.Info.Name);
```
minor nit to ignore: for consistency it could be First(w => **w**.Info.Name == Info.BodyName) (also in RulesetLoaded). Not changed by this PR but I assume .Single could be replaced by First which is faster. We check for single WithSpriteBodyInfo in RulesetLoaded so why should we rely on exception here for Armament?

rob-v commented on OpenRA.Mods.Common/Traits/Render/WithAttackAnimation.cs:
```diff
@@ -34,68 +34,87 @@ public class WithAttackAnimationInfo : ITraitInfo, Requires<WithSpriteBodyInfo>,
 		{
-			if (!string.IsNullOrEmpty(info.AttackSequence))
-				wsb.PlayCustomAnimation(self, info.AttackSequence, () => wsb.CancelCustomAnimation(self));
+			if (!IsTraitDisabled && wsb.IsTraitDisabled && !string.IsNullOrEmpty(Info.AttackSequence))
```
Is it correct to call wsb.PlayCustomAnimation when wsb.IsTraitDisabled and not enabled?

reaperrr replied:
No, good catch.

reaperrr commented:
Updated, replaced `.Single` with `.First` in the armament look-up.

Even ignoring performance, it is entirely possible that mods might want to 'share' an attack anim between two armaments (for example, dual-barreled units might need two armaments to fire from both barrels at the same time, imagine TS Wolverine firing visible projectiles), so being forced to give one of the armaments a different name just to avoid a crash is a bit inconvenient.

rob-v commented:
Found issues updated, looks good, but not tested (not clear for me how to test it ingame), enough reviewers, so only small 👍 from me.

reaperrr committed:
Rename WithAttackAnimation.BodyName to just .Body

Shorter and more consistent with Armament.Turret, WithTurretedAttackAnimation.Turret, WithSpriteBarrel.Armament etc.

reaperrr commented:
Rebased and updated. Dropped the Single->First commit again (see discussion on #13378).

@rob-v If you add
```
WithFacingSpriteBody@2:
	Name: body2
```

to RA V2RL or TS SMECH, they'll crash on bleed. With this PR, they'll show both bodies, but only the default one will be modified by the attack anim trait.

Additionally, you can test for regressions just by checking if V2RL aim/reload anim and TS Wolverine attack anim still work.

The conditional part could by tested by making `WithAttackAnimation` on one of the above actors require `rank-veteran > 1`, then the anim would only play when the unit ranks up by at least 1 level.

rob-v commented:
This is probably not related to this PR, when I add WithFacingSpriteBody to smech as you proposed, it fails with this error: 'Unit `smech` does not have a sequence named `idle`'

reaperrr commented:
> This is probably not related to this PR, when I add WithFacingSpriteBody to smech as you proposed, it fails with this error: 'Unit smech does not have a sequence named idle'

Ah yes, the default value for `Sequence` in sprite bodies is `idle`, which smech doesn't have, so for my example to work, you'd have to set `Sequence: stand` on that additional sprite body.

rob-v commented:
note - this `throw new YamlException("WithAttackAnimation needs exactly one sprite body with matching name.");` and other exceptions to be useful depends on this bug fix: #13372, otherwise these exceptions are not logged.

rob-v reviewed (approved):
👍

atlimit8 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/9faf73bd1ce3328cedf052ce59e84e19a47cfb37...74143e9b77b0e84b874e1b351f213ae92e44b8f2)
